Bump PyO3 and rust-numpy to 0.23.x

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit bumps the version of pyo3 and rust-numpy used by qiskit to the latest release 0.23. The largest change by volume of code is the deprecation of all the `*_bound()` methods. These are just warnings but they would be fatal to our CI so it needs to be updated. The larger functional change that required updating the code is the change in the traits around converting to Python objects. This actually found a bug in the target where we were not returning a proper instruction type for standard gates. This also opens up the opportunity to update our hashbrown version to 0.15.x, but we can't do that until the next rustworkx-core release.

### Details and comments

mtreinish committed:
Bump PyO3 and rust-numpy to 0.23.x

This commit bumps the version of pyo3 and rust-numpy used by qiskit to
the latest release 0.23. The largest change by volume of code is the
deprecation of all the `*_bound()` methods. These are just warnings but
they would be fatal to our CI so it needs to be updated. THe larger
functional change that required updating the code is the change in the
traits around converting to Python objects. This actually found a bug in
the target where we were not returning a proper instruction type for
standard gates. This also opens up the opportunity to update our
hashbrown version to 0.15.x, but we can't do that until the next
rustworkx-core release.

mtreinish committed:
Merge remote-tracking branch 'origin/main' into pyo3-bump-23

mtreinish committed:
Fix cache pygates build

mtreinish committed:
Fix cargo test

kevinhartman commented on crates/accelerate/src/consolidate_blocks.rs:
```diff
@@ -180,10 +180,11 @@ pub(crate) fn consolidate_blocks(
-                        .get_bound(py)
-                        .call1((array.to_object(py), py.None(), false))?;
+                let unitary_gate = UNITARY_GATE.get_bound(py).call1((
+                    <pyo3::Bound<'_, PyArray<Complex<f64>, ndarray::Dim<[usize; 2]>>> as Clone>::clone(&array).into_pyobject(py)?,
```
👀

kevinhartman replied:
Maybe this?

```suggestion
                    array.as_ref().into_pyobject(py)?,
```

mtreinish committed:
Fix impl of IntoPyObject for target's operation

mtreinish committed:
Fix impl of IntoPyObject for equivalence library circuit

kevinhartman commented on crates/accelerate/src/sabre/heuristic.rs:
```diff
@@ -159,9 +174,11 @@ impl DecayHeuristic {
-            .into_py(py))
+            .into_pyobject(py)?
+            .into_any()
+            .unbind())
```
Is this potentially interesting for these cases?

https://pyo3.rs/main/doc/pyo3/conversion/trait.intopyobjectext#method.into_py_any

mtreinish replied:
Good idea, I updated this in: https://github.com/Qiskit/qiskit/pull/13577/commits/dd1ce36d1bae71e381e25f10d825e249ea2cda00

kevinhartman commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -112,6 +95,26 @@ pub(crate) struct NormalOperation {
+
+impl<'py> IntoPyObject<'py> for &NormalOperation {
+    type Target = PyAny;
+    type Output = Bound<'py, Self::Target>;
```
From the migration guide, should this be a `Borrowed` instead?

https://pyo3.rs/v0.23.3/migration.html#intopyobject-manual-implementation

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/13577/commits/17210ee76c8c4395b69d709184d5a286c069fe60

kevinhartman commented on crates/circuit/src/operations.rs:
```diff
@@ -72,25 +73,35 @@ impl<'py> FromPyObject<'py> for Param {
-        }
-    }
-}
+//impl<'py> IntoPyObject<'py> for Param {
```
👀

mtreinish replied:
Oops this was leftover from me playing around before I derived the implementation

mtreinish replied:
Removed in: https://github.com/Qiskit/qiskit/pull/13577/commits/c32f590d0a106a0d96601918244d22c383f196c4

kevinhartman reviewed (commented):
Thanks for doing this! I've done just one quick pass for now, and will unassign myself since I'm unlikely to have another look before I'm out on vacation.

mtreinish committed:
Merge remote-tracking branch 'origin/main' into pyo3-bump-23

mtreinish committed:
Remove commented out code

mtreinish committed:
Fix commutation library conversion

mtreinish committed:
Fix conversion of qasm3 py register type

mtreinish committed:
Merge remote-tracking branch 'origin/main' into pyo3-bump-23

mtreinish committed:
Use into_py_any()

mtreinish committed:
Merge remote-tracking branch 'origin/main' into pyo3-bump-23

mtreinish committed:
Use borrowed instead of cloning py bound for NormalOperation

mtreinish committed:
Fix unitary synthesis failure

In pyo3 0.23 the behavior of the default conversion for a `SmallVec<u8>`
and similar arrays of u8 default to bytes objects when converted to
python. This was documented as an api change in the release, however it
was cauing the unitary synthesis test to fail because when we were
evaluating whether the synthesis was in terms of the natural direction
of the 2q gate on the backend it was evaluating `[0, 1] == b"\x00\x11"`
which evaluates to `False` and the pass flipped the direction of the
2q gate. This was causing the test failure because all the 2q gates
which were correctly directed were getting incorrectly flipped. This
commit fixes this by manually creating a pylist so the comparisons work
as expected.

mtreinish committed:
Fix cargo fmt

kevinhartman commented on crates/accelerate/src/basis/basis_translator/compose_transforms.rs:
```diff
@@ -106,7 +106,7 @@ pub(super) fn compose_transforms<'a>(
                     let param_mapping: HashMap<ParameterUuid, Param> = equiv_params
                         .iter()
-                        .map(|x| ParameterUuid::from_parameter(x.to_object(py).bind(py)))
+                        .map(|x| ParameterUuid::from_parameter(&x.into_pyobject(py).unwrap()))
```
For others, the `unwrap` is needed here because PyO3's `IntoPyObject::into_pyobject` now returns a `Result<T, E>` type. The error type is `PyErr` in this case (dictated by the `#[derive(IntoPyObjectRef)]` attribute on `Param`).

jakelishman replied:
It's not clear to me why the `Param` needs converting into a Python object at all here.  If it isn't a `Param::Parameter` variant, the whole thing will fail, so it seems like we could have sorted out that typing a while prior.

kevinhartman commented on crates/accelerate/src/commutation_checker.rs:
```diff
@@ -725,6 +721,31 @@ pub enum CommutationLibraryEntry {
+            CommutationLibraryEntry::Commutes(b) => {
+                let temp = b.into_pyobject(py)?;
+                <pyo3::Bound<'_, PyBool> as Clone>::clone(&temp).into_any()
+            }
```
I think this would be the expected way to do this. The impl of `IntoPyObject` for `bool` gives us a `Borrowed<PyBool>`, and then by calling trait method `BoundObject::into_bound` on that, we get an owned `Bound<PyBool>`.

```suggestion
            CommutationLibraryEntry::Commutes(b) => b.into_pyobject(py)?.into_bound().into_any(),
```

kevinhartman commented on crates/accelerate/src/sparse_observable.rs:
```diff
@@ -1315,14 +1318,21 @@ impl SparseObservable {
+                    .to_term()
+                    .into_pyobject(py)?
+                    .into_any()
+                    .unbind())
```
Looks like you can use `into_py_any(py)` here if you `use pyo3::IntoPyObjectExt;`.

mtreinish replied:
I knew I would miss at least one of these when I updated this

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/13577/commits/389f6e7a4fcd92e11e00b57a57d57471fee499b1

kevinhartman commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -57,30 +58,12 @@ type GateMapState = Vec<(String, Vec<(Option<Qargs>, Option<InstructionPropertie
 /// Represents a Qiskit `Gate` object or a Variadic instruction.
 /// Keeps a reference to its Python instance for caching purposes.
-#[derive(FromPyObject, Debug, Clone)]
+#[derive(FromPyObject, Debug, Clone, IntoPyObjectRef)]
```
Should we also derive `IntoPyObject`?

mtreinish replied:
Probably, I probably didn't do it because nothing used it, I'll add this. There might havee been a reason I didn't add it, but I'll find out when I try this (and will report back if it can't be added).

mtreinish replied:
Ah, so trying this locally adding the definition confuses the return of `self._gate_name_map[x].into_py_object()` below because it tries to do it with an owned version instead of the ref.

jakelishman replied:
I imagine that's expected / intended - I'd assume the spelling would need to become `(&self._gate_name_map[x]).into_py_object()`.  I think in general `t.into_py()` goes to `t.into_py_object()` and `t.to_object()` goes to `(&t).into_py_object()`?

kevinhartman commented on crates/circuit/src/interner.rs:
```diff
@@ -206,7 +206,8 @@ mod test {
         assert_eq!(interner.get_default(), interner.get_default());
-        assert_eq!(interner.get(interner.get_default()), &[]);
+        let res: &[u32] = &[];
+        assert_eq!(interner.get(interner.get_default()), res);
```
What's going on here?

mtreinish replied:
TBH I can't remember. It's either the compiler started complaining here and suggested I do this to make it work, or I was debugging something and did this to do `println!("{:?}", res)` before the assertion and forgot to change it back but removed the print. I'll try reverting this and see what happens.

mtreinish replied:
After trying it locally I remember now. The compiler was complaining it can't infer the type of `&[]` anymore so I used the res var to define it as a `&[u32]`.

jakelishman replied:
This seems weird that the PyO3 update would cause the inference here to suddenly fail.  The types of `Interner` and `get_default()` shouldn't have changed.

kevinhartman commented on crates/qasm2/src/bytecode.rs:
```diff
@@ -211,89 +211,108 @@ pub enum InternalBytecode {
 }
 
-impl IntoPy<Bytecode> for InternalBytecode {
+impl<'py> IntoPyObject<'py> for InternalBytecode {
```
Is this not derivable because the enum variants have named fields?

mtreinish replied:
Yeah when I did the derive approach it didn't work, I don't remember the exact details because I worked on this > 1 month ago. But I had to go this route.

jakelishman replied:
Using the semantics of the derive macros would substantially change how the Rust/Python boundary layer is handled here.  Even if PyO3 didn't error, the types would be entirely different to what Python is expecting.

kevinhartman reviewed (commented):
Just a few suggestions, otherwise this looks ready to me. Thanks!

mtreinish committed:
Apply suggestions from code review

Co-authored-by: Kevin Hartman <kevin@hart.mn>

mtreinish committed:
Fix compilation error

mtreinish committed:
Use into_py_any in missed spot

mtreinish committed:
Fix lint

mtreinish committed:
Fix cargo fmt

mtreinish committed:
Merge branch 'main' into pyo3-bump-23

jakelishman commented on crates/accelerate/src/error_map.rs:
```diff
@@ -101,14 +102,19 @@ impl ErrorMap {
+        &self,
+        py: Python,
+        key: [PhysicalQubit; 2],
+        default: Option<PyObject>,
+    ) -> PyResult<PyObject> {
+        Ok(match self.error_map.get(&key).copied() {
+            Some(val) => val.into_py_any(py)?,
             None => match default {
                 Some(val) => val,
                 None => py.None(),
             },
```
Seems weird that we do `default: Option<PyObject>` when we just recreate the `None` singleton anyway, but that was pre-existing.

jakelishman commented on crates/accelerate/src/euler_one_qubit_decomposer.rs:
```diff
@@ -109,11 +110,13 @@ impl OneQubitGateSequence {
         match idx.with_len(self.gates.len())? {
-            SequenceIndex::Int(idx) => Ok(self.gates[idx].to_object(py)),
-            indices => Ok(PyList::new_bound(
+            SequenceIndex::Int(idx) => Ok(self.gates[idx].clone().into_py_any(py)?),
```
The `clone` and move seems weird here.  Does it work with `(&self.gates[idx]).into_py_any(py)` to do it implicitly (and potentially make it easier to elide the clone/move)?

jakelishman replied:
I solved this by implementing `IntoPyObject for &T` for all our local `T: PyClass + Clone` (since I can't do the blanket impl before of coherence rules).  I might ask over on PyO3 if it's something that should have a blanket impl, be part of the `pyclass` generation, or something else.

jakelishman commented on crates/accelerate/src/euler_one_qubit_decomposer.rs:
```diff
@@ -109,11 +110,13 @@ impl OneQubitGateSequence {
-            )
+                indices
+                    .iter()
+                    .map(|pos| self.gates[pos].clone().into_pyobject(py).unwrap()),
```
Same here with the `clone`.

jakelishman commented on crates/accelerate/src/euler_one_qubit_decomposer.rs:
```diff
@@ -712,12 +715,8 @@ impl EulerBasis {
-            (PyString::new_bound(py, self.as_str()),),
-        )
-            .into_py(py)
+    fn __reduce__(&self, py: Python) -> PyResult<Py<PyAny>> {
+        (py.get_type::<Self>(), (PyString::new(py, self.as_str()),)).into_py_any(py)
```
This `__reduce__` is actually a fancy way of spelling `__getnewargs__`, but I think I might have been the culprit here.

jakelishman commented on crates/accelerate/src/nlayout.rs:
```diff
@@ -123,12 +115,11 @@ impl NLayout {
-            (self.virt_to_phys.to_object(py),),
+        (
+            py.get_type::<Self>().getattr("from_virtual_to_physical")?,
+            (self.virt_to_phys.clone().into_pyobject(py).unwrap(),),
```
```suggestion
            (self.virt_to_phys.clone().into_pyobject(py)?,),
```
since we're already in a `PyResult`-returning function.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/13577/commits/0577f48a7814455551ec2cda3540a9d1a7ad9847

jakelishman commented on crates/accelerate/src/nlayout.rs:
```diff
@@ -143,8 +134,8 @@ impl NLayout {
     #[pyo3(text_signature = "(self, /)")]
-    fn layout_mapping(&self, py: Python<'_>) -> Py<PyList> {
-        PyList::new_bound(py, self.iter_virtual()).into()
+    fn layout_mapping(&self, py: Python<'_>) -> PyResult<Py<PyList>> {
+        PyList::new(py, self.iter_virtual()).map(|x| x.unbind())
```
Maybe better to change the return type to `PyResult<Bound<'py, PyList>>` rather than add an explicit unbind?

jakelishman commented on crates/accelerate/src/sparse_observable.rs:
```diff
@@ -1385,15 +1391,20 @@ impl SparseObservable {
                 &slf_.borrow(),
                 Complex64::new(2.0, 0.0),
             )
-            .into_py(py));
+            .into_pyobject(py)?
+            .into_any()
+            .unbind());
```
Spiritually, I feel like these magic methods should have just started reuning `Bound<...>` instead of `Py<...>`, but we don't need to change it again in this PR.

jakelishman commented on crates/accelerate/src/sparse_observable.rs:
```diff
@@ -1547,7 +1579,7 @@ impl SparseObservable {
     /// Examples:
     ///
     ///     .. code-block:: python
-    ///         
+    ///
```
Huh, weird to me that rustfmt doesn't tidy these up in comments.

jakelishman commented on crates/accelerate/src/two_qubit_decompose.rs:
```diff
@@ -1295,11 +1297,16 @@ impl TwoQubitGateSequence {
         match idx.with_len(self.gates.len())? {
-            SequenceIndex::Int(idx) => Ok(self.gates[idx].to_object(py)),
-            indices => Ok(PyList::new_bound(
+            SequenceIndex::Int(idx) => {
+                let item = &self.gates[idx];
+                (item.0, PyList::new(py, &item.1)?, PyList::new(py, &item.2)?).into_py_any(py)
+            }
+            indices => Ok(PyList::new(
```
Is this working around the `[u8]`-is-now-`bytes` conversion trait weirdness?

mtreinish replied:
Yes this is how I had to workaround that. See: https://github.com/Qiskit/qiskit/pull/13577/commits/c59dc9c071d002fcc5288b0ba3c55a9979a19b3e where I made this change

jakelishman commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -910,9 +905,10 @@ impl DAGCircuit {
                 py,
-                r#"
+                std::ffi::CString::new(
+                    r#"
```
If only we were MSRV 1.77 and could just write `cr#"`...

jakelishman commented on crates/circuit/src/dag_node.rs:
```diff
@@ -249,21 +252,17 @@ impl DAGOpNode {
-            state,
-        )
-            .into_py(py))
+        let temp = (
+            slf.instruction.get_operation(py)?,
+            &slf.instruction.qubits,
+            &slf.instruction.clbits,
+        );
```
I wouldn't mind calling `temp` something like `new_args` - the `__reduce__` tuple is complex enough as it is.

jakelishman commented on crates/circuit/src/lib.rs:
```diff
@@ -33,7 +33,7 @@ use pyo3::prelude::*;
 pub type BitType = u32;
-#[derive(Copy, Clone, Debug, Hash, Ord, PartialOrd, Eq, PartialEq, FromPyObject)]
+#[derive(Copy, Clone, Debug, Hash, Ord, PartialOrd, Eq, PartialEq, FromPyObject, IntoPyObject)]
 pub struct Qubit(pub BitType);
```
This looks like a behaviour change - `Qubit` and `Clbit` didn't previously have `IntoPyObject` defined, I think?

mtreinish replied:
Removed in: https://github.com/Qiskit/qiskit/pull/13577/commits/132de0ad23929c2f8ff3c1296872eec5a1b075be

jakelishman commented on crates/circuit/src/operations.rs:
```diff
@@ -688,9 +663,12 @@ impl StandardGate {
+        params: Vec<Param>,
+    ) -> Option<Bound<'py, PyArray2<Complex64>>> {
+        self.matrix(&params).map(|x| x.into_pyarray(py))
     }
```
I don't think the lifetime bound on `self` needs to live as long as `'py`?  Not that I think it matters, since this method is marked "just for testing".

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/13577/commits/a2813401119cd80360cfb6891ea30c41525d60f8

jakelishman commented on crates/circuit/src/packed_instruction.rs:
```diff
@@ -111,7 +111,7 @@ unsafe impl ::bytemuck::NoUninit for PackedOperationType {}
-#[derive(Debug)]
+#[derive(Debug, IntoPyObject, IntoPyObjectRef)]
 #[repr(transparent)]
 pub struct PackedOperation(usize);
```
This really feels wrong - `PackedOperation` definitely should not ever be exposed to Python.  Aside from the data being completely meaningless to Python, implementing `IntoPyObjectRef` for it allows the holder to `Clone` the inner `usize`, which might represent ownership of a `Box` pointer.

mtreinish replied:
Removed in: https://github.com/Qiskit/qiskit/pull/13577/commits/132de0ad23929c2f8ff3c1296872eec5a1b075be

jakelishman commented on crates/circuit/src/parameter_table.rs:
```diff
@@ -10,7 +10,7 @@
 // that they have been altered from the originals.
 
-use std::cell::OnceCell;
+use std::sync::OnceLock;
```
Is this because of free-threading support in PyO3, or a separate behavioral change?

mtreinish replied:
Yeah, because we need the members of a pyclass need to support be used in a threading context this was needed to make `CircuitData` as a pyclass work with pyo3 0.23

jakelishman commented on crates/qasm2/src/expr.rs:
```diff
@@ -146,57 +146,76 @@ pub enum Expr {
-            Expr::Constant(value) => bytecode::ExprConstant { value }.into_py(py),
-            Expr::Parameter(index) => bytecode::ExprArgument { index }.into_py(py),
+impl<'py> IntoPyObject<'py> for Expr {
+    type Target = PyAny; // the Python type
+    type Output = Bound<'py, Self::Target>; // in most cases this will be `Bound`
```
lol, this isn't the first time I've seen the copy/paste comments from the guide

jakelishman commented on crates/qasm2/src/lex.rs:
```diff
@@ -337,7 +337,7 @@ pub struct TokenStream {
     pub filename: std::ffi::OsString,
     strict: bool,
-    source: Box<dyn std::io::BufRead + Send>,
+    source: Box<dyn std::io::BufRead + Send + Sync>,
```
Is the `Sync` bound now required because of the free-threading support?

mtreinish replied:
Yeah, it was a requirement that the members of a pyclass were all Sync

jakelishman commented on crates/qasm3/src/circuit.rs:
```diff
@@ -86,30 +83,38 @@ pub struct PyGate {
         num_qubits: usize,
-    ) -> Self {
+    ) -> Self
+    where
+        <T as pyo3::IntoPyObject<'py>>::Error: std::fmt::Debug,
```
This bound feels pretty questionable - should we just make the constructor fallible and have callers bubble it up?

mtreinish replied:
Removed in: https://github.com/Qiskit/qiskit/pull/13577/commits/0577f48a7814455551ec2cda3540a9d1a7ad9847

jakelishman commented on crates/qasm3/src/circuit.rs:
```diff
@@ -86,30 +83,38 @@ pub struct PyGate {
     {
-        let args = args.into_py(py);
-        let received_num_params = args.bind(py).len();
+        let args = args.into_pyobject(py).unwrap();
```
This function already returns `PyResult`, so no need for the crazy trait bounds and `unwrap`, I think?

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/13577/commits/0577f48a7814455551ec2cda3540a9d1a7ad9847

jakelishman commented on crates/qasm3/src/circuit.rs:
```diff
@@ -264,24 +270,36 @@ impl PyCircuitModule {
+                qubits.into_pyobject(py)?,
+                clbits.into_pyobject(py)?,
+            ),
+        )
```
I think this could be `IntoPyObjectExt::into_pyobject_or_pyerr`, which would automatically use the trait bounds from the `IntoPyObject::Error` associated type to avoid needing to spell them out again.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/13577/commits/0577f48a7814455551ec2cda3540a9d1a7ad9847

jakelishman reviewed (commented):
Thanks for doing this.  I think there's a fair number of function returns that spiritually ought to swap from `Py<PyAny>` (or the `PyObject` spelling) into `Bound<PyAny>` - I think making that change was one of the driving reasons to make the trait change in the first place.  We have a lot of explicit `unbind`s in places where we're returning the object itself to Python space, not hiding it in a Rust object that needs to be GIL-independent (could be `into_bound_py_any` in many cases?).  No need to go hunting for them, though, for this PR.

(Side note: it's weird to me that `into_bound_py_any` was chosen, rather than having `into_py_any` getting called `into_unbound_py_any`, given the shift of focus away from `Py` in those conversions.)

I feel like there's at least a couple of places I've written that does the same thing as `IntoPyObjectExt::into_pyobject_or_pyerr`, but I don't really remember them off the top of my head.

It's most likely lack of experience, but I've got no intuition for what the `IntoPyObject` derive macro will do to enums with compound variants.  Personally, I'm not sure I like the derive macros for record structs much at all - turning stuff into `dict` feels icky - but that's just a knee-jerk reaction right now and I'll see what I think with more experience.

I didn't manage to 100% convince myself that all the `Target` changes produce visibly equivalent code, so I'm pretty much banking on the tests there.

mtreinish committed:
Remove overeager trait derives

mtreinish committed:
Pull in latest pyo3 bugfix release

mtreinish committed:
Fix error handling

mtreinish committed:
Fix lifetime for _to_matrix()

mtreinish committed:
Merge branch 'main' into pyo3-bump-23

jakelishman commented:
I'm sorting out the merge conflicts and looking into fixing up a couple of the comments I left in the last review, then I'll merge.

jakelishman committed:
Merge remote-tracking branch 'ibm/main' into pyo3-bump-23

jakelishman committed:
Implement `IntoPyObject` for `&StandardGate`

This makes it more ergonomic to use the blanket implementations of
`IntoPyObject` on ad-hoc structures like `(&T1, &T2)` when one of the
contained types is a `StandardGate`.

jakelishman committed:
Implement `IntoPyObject` for all `&T` where `T: PyClass + Copy`

jakelishman reviewed (approved):
I'm happy with this now, though somebody else ought to sign off on my commit cd99d9fc88e28125374a2ef90ea9ddb87b915ab5.

mtreinish commented:
[That](https://github.com/Qiskit/qiskit/pull/13577/commits/cd99d9fc88e28125374a2ef90ea9ddb87b915ab5) looks good to me, so I went ahead and added it to merge queue.
